src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorDynamicFileInfoProvider.cs

```diff
@@ -0,0 +1,239 @@
+
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (project.WorkspaceProject == null)
+            {
+                // Don't bother if this isn't assocated with a project.
+                return;
+            }
+
+            // There's a possible race condition here where we're processing an update
+            // and the project is getting unloaded. So if we don't find an entry we can
+            // just ignore it.
+            var key = new Key(project.WorkspaceProject.Id, project.WorkspaceProject.FilePath, document.FilePath);
+            if (_entries.TryGetValue(key, out var entry))
+            {
+                var updated = false;
+                lock (entry.Lock)
+                {
+                    if (entry.Current.TextLoader is GeneratedDocumentTextLoader)
+                    {
+                        updated = true;
+                        entry.Current = CreateEmptyInfo(key);
+                    }
+                }
+
+                if (updated)
+                {
+                    Updated?.Invoke(this, document.FilePath);
+                }
+            }
+        }
+
+        public Task<DynamicFileInfo> GetDynamicFileInfoAsync(ProjectId projectId, string projectFilePath, string filePath, CancellationToken cancellationToken)
+        {
+            if (projectFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(projectFilePath));
+            }
+
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            var key = new Key(projectId, projectFilePath, filePath);
+            var entry = _entries.GetOrAdd(key, _createEmptyEntry);
+            return Task.FromResult(entry.Current);
+        }
+
+        public Task RemoveDynamicFileInfoAsync(ProjectId projectId, string projectFilePath, string filePath, CancellationToken cancellationToken)
+        {
+            if (projectFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(projectFilePath));
+            }
+
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            var key = new Key(projectId, projectFilePath, filePath);
+            _entries.TryRemove(key, out var entry);
+            return Task.CompletedTask;
+        }
+
+        private DynamicFileInfo CreateEmptyInfo(Key key)
+        {
+            var filename = Path.ChangeExtension(key.FilePath, ".g.cs");
+            var textLoader = new EmptyTextLoader(filename);
+            return new DynamicFileInfo(filename, SourceCodeKind.Regular, textLoader, _factory.CreateEmpty());
+        }
+
+        private DynamicFileInfo CreateInfo(Key key, DocumentSnapshot document)
+        {
+            var filename = Path.ChangeExtension(key.FilePath, ".g.cs");
```

Comment:
is `.g.cs` a common pattern for dynamic file infos like this? Wondering if we should be more explicit in differentiating file names to avoid collisions.

Reply:
yeah, I think elsewhere we've used `.g.chtml.cs` to make it distinct. That's a good idea.